modules/gem5/_posts/2000-03-01-assignment1.md

```diff
@@ -130,7 +121,44 @@ Use `4GHz` as the clock frequency.
 
 In your report, answer the same questions after simulation supported with data. A complete set of simualtion data for this step should include **6 configurations** (2 options for CPU model * 3 options for memory model).
 
-### Step III: General questions
+### Step III: Using different compiler optimizations
+
+In this step, you'll be using different compiler optimizations to run the same
+matrix multiplication program. Compiler optimization flags are options that can
+be used to improve the *performance* of the program at the expense of
+compilation time and the ability to debug the program. The default version of
+the matrix multiply program is compiled with the flag `-O2`:
+
+```sh
+g++ -o mm mm.cpp -static -O2
+```
+
+We have compiled binaries with flags `-O0` and `-O3`. The former reduces the
+compilation time and makes debugging produce expected results. The latter
+optimizes the code to the maximum possible limit. This *usually* means that
+there are less number of instructions in the more optimized version of the
+program. Let's look at the following computation: $z = \alpha x + y$.
+This operation first multiplies $\alpha$ and $x$, and then adds $y$ to the
+result. Assume that x, y and z are all float. In modern processors,
+these two steps are usually combined to do a `multiply and accumulate`
+operation in a single step. If the compiler detects such computations, it'll
+try to optimize this from two `ld` and one `st` to a MAC operation like
+`fmadd.d` instruction.
+When we look at the assembly, it generates an `fmadd.d rd, rs1, rs2, rs3`.
```

Comment:
It could be useful to link to a saved example on [godbolt](https://godbolt.org/)

Reply:
I'll do it once i am done with assignment 3. i need to spend some time with getting the correct results.